q2_cutadapt/tests/test_demux.py

```diff
@@ -6,17 +6,185 @@
 # The full license is in the file LICENSE, distributed with this software.
 # ----------------------------------------------------------------------------
 
+import gzip
+import os
+import pathlib
+import shutil
 import unittest
 
+import pandas as pd
+
+from q2_cutadapt._demux import (_build_demux_command, _rename_files,
+                                _write_metadata_yaml_in_results,
+                                _write_manifest_in_results,
+                                _write_empty_fastq_to_mux_barcode_in_seq_fmt)
+from q2_types.multiplexed_sequences import (
+    MultiplexedSingleEndBarcodeInSequenceDirFmt)
+from q2_types.per_sample_sequences import (
+    SingleLanePerSampleSingleEndFastqDirFmt, FastqGzFormat)
+from qiime2 import Artifact, MetadataCategory
+from qiime2.util import redirected_stdio
 from qiime2.plugin.testing import TestPluginBase
 
 
 class TestDemuxSingle(TestPluginBase):
     package = 'q2_cutadapt.tests'
 
-    def test_nothing(self):
-        # This is just so that the tests won't error when run
-        self.assertTrue(True)
+    def setUp(self):
+        super().setUp()
+        self.demux_single_fn = self.plugin.methods['demux_single']
+
+        muxed_sequences_fp = self.get_data_path('forward.fastq.gz')
+        self.muxed_sequences = Artifact.import_data(
+            'MultiplexedSingleEndBarcodeInSequence', muxed_sequences_fp)
+
+    def test_typical(self):
+        metadata = MetadataCategory(pd.Series(['AAAA', 'CCCC'],
+                                    index=['sample_a', 'sample_b'],
+                                    name='Barcode'))
+
+        with redirected_stdio(stderr=os.devnull):
+            obs_demuxed_art, obs_untrimmed_art = \
+                self.demux_single_fn(self.muxed_sequences, metadata)
+
+        obs_demuxed = obs_demuxed_art.view(
+            SingleLanePerSampleSingleEndFastqDirFmt)
+        exp_samples_and_barcodes = metadata.to_series().iteritems()
+        obs_demuxed_seqs = obs_demuxed.sequences.iter_views(FastqGzFormat)
+        zipped = zip(exp_samples_and_barcodes, obs_demuxed_seqs)
+        for (sample_id, barcode), (filename, _) in zipped:
+            filename = str(filename)
+            self.assertTrue(sample_id in filename)
+            self.assertTrue(barcode in filename)
+
+        obs_untrimmed = obs_untrimmed_art.view(
+            MultiplexedSingleEndBarcodeInSequenceDirFmt)
+        obs_untrimmed = obs_untrimmed.file.view(FastqGzFormat)
+        obs_untrimmed = gzip.decompress(obs_untrimmed.path.read_bytes())
+        self.assertEqual(b'@id6\nGGGGACGTACGT\n+\nzzzzzzzzzzzz\n',
+                         obs_untrimmed)
+
```

Comment:
Add a test for having an extra barcode in the metadata column

Reply:
Also maybe a test for variable length barcodes?

Reply:
Sounds good! Also need to expose the error tolerance, will add some tests for that, too.